src/appleseed/renderer/meta/tests/test_projectfilewriter.cpp

```diff
@@ -60,301 +62,230 @@ TEST_SUITE(Renderer_Modeling_Project_ProjectFileWriter)
         {
-            filesystem::remove_all(m_base_output);
-            filesystem::remove_all(m_alternate_output);
+            remove_all(m_output_directory);
 
-            filesystem::create_directory(m_base_output);
-            filesystem::create_directory(m_alternate_output);
+            // On Windows, the create_directory() call below will fail with an Access Denied error
+            // if a File Explorer window was opened in the output directory that we just deleted.
+            // A small pause solves the problem.
+            sleep(50);
+
+            create_directory(m_output_directory);
         }
 
         void create_project()
         {
             m_project = ProjectFactory::create("project");
             m_project->set_scene(SceneFactory::create());
 
-            m_project->set_path((m_base_output / "input.appleseed").string().c_str());
-            m_project->search_paths().set_root_path(m_base_output.string());
+            m_project->set_path((m_output_directory / "project.appleseed").string().c_str());
+            m_project->search_paths().set_root_path(m_output_directory.string());
         }
 
-        void create_texture_entity(const string& filepath)
+        void create_assembly()
         {
-            ParamArray params;
-            params.insert("filename", filepath);
-            params.insert("color_space", "linear_rgb");
-
-            m_project->get_scene()->textures().insert(
-                DiskTexture2dFactory().create(
-                    "texture",
-                    params,
-                    m_project->search_paths()));
+            m_project->get_scene()->assemblies().insert(AssemblyFactory().create("assembly"));
         }
 
-        void create_texture_file(const filesystem::path& filepath)
+        Assembly* get_assembly()
         {
-            const filesystem::path fullpath = m_base_output / filepath;
-            filesystem::create_directories(fullpath.parent_path());
-            filesystem::copy_file("unit tests/inputs/test_projectfilewriter_texture.png", fullpath);
+            return m_project->get_scene()->assemblies().get_by_name("assembly");
         }
 
-        void create_geometry_file(const filesystem::path& filepath)
+        template <typename T>
+        void create_mesh_object(const char* object_name, const T& filename)
         {
-            const filesystem::path fullpath = m_base_output / filepath;
-            filesystem::create_directories(fullpath.parent_path());
-            filesystem::copy_file("unit tests/inputs/test_projectfilewriter_object.obj", fullpath);
+            get_assembly()->objects().insert(
+                auto_release_ptr<Object>(
+                    MeshObjectFactory::create(
+                        object_name,
+                        ParamArray().insert("filename", filename))));
         }
 
-        string get_texture_entity_filepath() const
+        template <>
```

Comment:
Interesting. I started with an overload but then thought a specialization would be clearer. VS does not complain, obviously. Will fix now.